### ISSUE TITLE
Syntax highlighting for text/template and html/template

### ISSUE THREAD

**Issue by andradei** [microsoft/vscode-go#464]
It would be great to have `html/template` extend VS Code's HTML syntax highlighting and add support for Go's default tags `{{` and `}}` and the logic in between them.

**Comment by MikeK123** [comment]
This feature would be much appreciated.

**Comment by tbroyer** [comment]
FWIW, just found out about https://marketplace.visualstudio.com/items?itemName=casualjim.gotemplate
Not tried it yet.

**Comment by ramya-rao-a** [comment]
Is this feature request for having syntax highlighting for inline template strings in go files? Or for having syntax highlighting for files that contain the html or text template with Go's tags `{{ }}` ?

Also is this request the same as #228 or different?

Both this issue and #228 are getting a lot of upvotes, I am trying to understand the expected behaviour.

**Comment by andradei** [comment]
@ramya-rao-a I think #228 is referring to having things like auto-completion and go-to support when writing/editing templates.

This one would be for syntax highlighting of templates, specifically for code between `{{ }}`.

So for example: for `html/template` it could be an extension of the default html syntax highlighting for VSCode, which can also highlight code inside `{{ }}` on `.html` files.

It would also be nice to have support for file extensions like:

- .gohtml
- .template
- .tmpl
- .html

etc.

**Comment by ramya-rao-a** [comment]
Thanks @andradei 
In that case the extension that @tbroyer mentioned seem to be doing the needful.

@alexcrownus, @gertcuykens , @jimmykuu , @krmcbride, @xeleven, @MikeK123 , @guanxi55nba, @tampajohn (Folks who have upvoted this issue) Your thoughts on the extension? https://marketplace.visualstudio.com/items?itemName=casualjim.gotemplate ?

**Comment by gertcuykens** [comment]
The extension should be considered a temporary solution because it uses just custom syntax highlighting. For example the vscode html engine does allot more then just that.

To explain it more simply just rename your `.go` file that contains your template with let say 90% html code to `.html` Notice the much more in depth syntax highlighting + snippets + js support and every other build in feature what should be made available for a permanent consistent solution. Meaning updates to the build in html engine only need to be done in one place for all template languages either go or any other language like php or what ever.

So conclusion find a common ground and make sure renaming a go or php or ruby or python file to `.html` is consistent regarding html inside all the template languages and not depending on plugins that do the same thing over and over again.

**Comment by josephspurrier** [comment]
I'd also like to have go template formatting that supports the handlebars. For instance, I want the file to format like this:

```html
{{define "title"}}{{end}}
{{define "head"}}{{end}}
{{define "content"}}
	<div class="page-header">
		<h1>{{template "title" .}}</h1>
	</div>
	<p>This is the default template.</p>
	{{template "footer" .}}
{{end}}
{{define "foot"}}{{end}}
```

But when it formats on save, it looks like this:

```html
{{define "title"}}{{end}} {{define "head"}}{{end}} {{define "content"}}
<div class="page-header">
  <h1>{{template "title" .}}</h1>
</div>
<p>This is the default template.</p>
{{template "footer" .}} {{end}} {{define "foot"}}{{end}}
```

Would this be something that could go into the core of visual studio or something that a plugin would need to add functionality for?

Let me know if you want me to put this in a separate ticket.

**Comment by ramya-rao-a** [comment]
@josephspurrier We support the use of either `goreturns`, `goimports` or `gofmt` as formatting tools in the Go extension and do nothing extra.

So what is needed is a tool that does formatting like the above tools and in addition also format the html like you mentioned. This way you get the template formatting not just in VS Code but also for any other editor or even an user who can use the tool independent of an editor

**Comment by HintikkaKimmo** [comment]
@ramya-rao-a I think support for html templating languages is much more general than gohtml. VScode is missing similar feature for Django & Jinja templates as well and for handlebars what @josephspurrier is asking. 

gotemplate-syntex does part of what I am looking for with syntax highlighting, but it has no knowledge of my go variables or does not even know how to auto close `{{` signs. 

What I would like have as feature set:

1. gohtml, g.html template highlighting
2. Autocomplete for gohtml syntax and go variables used. 
3. Goto functionality allowing you to jump from .go file to referred template. 

Someone, please augment my idea as required. As said above similar functionality is missing from other templating engines as well from vscode.

**Comment by alexmironof** [comment]
And what about autoformat template files>

**Comment by pavelbinar** [comment]
Hi @ramya-rao-a 
any chance to get this feature to your Iteration Plan for 2018?
Thanks.

**Comment by ramya-rao-a** [comment]
@pavelbinar Which iteration plan are you referring to?

**Comment by pavelbinar** [comment]
@ramya-rao-a Sorry, if that was not clear.
The point of my question was: Do you plan to work on this feature any time this year?

**Comment by ramya-rao-a** [comment]
Hopefully yes :) But I havent gotten started on it yet. I'll do some initial investigations and add my findings here in the next month or so

**Comment by eaglebush** [comment]
It can certainly help for go templates to have syntax highlighting or to turn on linting. For instance, I have a dilemma of using the commonly used file extension for template file, which is .tmpl, but because VSCode does not recognize it, the code is not highlighted, or with .html or just .htm. Some problems popup, but it does not interfere with the code compiling. It is just that an annoying thing. Please see the screenshot.

![templateproblems](https://user-images.githubusercontent.com/32386756/52923428-c5cf0080-3362-11e9-8211-a58b227e7370.JPG)

Which is the result of this:

![templateproblems2](https://user-images.githubusercontent.com/32386756/52923431-c9628780-3362-11e9-8f17-fa8d22ebcbde.JPG)

**Comment by Luummii** [comment]
It does its job quite well for files with the extension .tmp
https://marketplace.visualstudio.com/items?itemName=ronnidc.nunjucks

**Comment by r03** [comment]
It would also be nice to have syntax color for inline template strings.
I only need {{ }} to be in a different color. (like %s is also colored in a string)

![image](https://user-images.githubusercontent.com/10061147/69726395-c631ce80-1120-11ea-9d5c-383478c79f4d.png)

**Comment by ramya-rao-a** [comment]
Hello all,

The feature being requested in this issue is syntax highlighting for html template code in Go files. Language features for such code like code completion and navigation is being discussed in https://github.com/microsoft/vscode-go/issues/228

I do see the merits of this feature, although I am afraid I don't believe this extension is the right place for it anymore.

As VS Code and this extension has gained popularity, we have seen a gradual scope creep over the past two years. Being a sole maintainer of this project, it is getting a little hard to support all incoming requests and maintaining the ever growing set of features. 

Moreover, this extension does not own or maintain the grammar that powers the syntax highlighting on Go files which comes out of the box from VS Code. This was previously based on https://github.com/atom/language-go and now moving towards https://github.com/jeff-hykin/better-go-syntax/ as being discussed in https://github.com/microsoft/vscode/issues/82549

Therefore, it makes sense to have the grammar to support html template come from the same source or at the very least a separate extension.

Another alternative would be to make this feature request in the upstream VS Code repo. It might be labelled as `extension-candidate` meaning that this feature is better implemented as a separate extension in which case one can ask for pointers on how to go about doing that, perhaps based on the jsx support in VS Code. 

My apologies and thanks for your patience.

**Comment by Kissaki** [comment]
Thank you very much for your efforts and the detailed explanation including possible alternatives and future prospects. Top notch!
